drivers/sensor/mc3419/mc3419.c

```diff
@@ -0,0 +1,302 @@
+{
+	double value = 0.0;
+
+	*raw_data = sys_le16_to_cpu(*raw_data);
+	value = (double)(*raw_data) * sensitivity *
+			 SENSOR_GRAVITY_DOUBLE / 1000;
+
+	val->val1 = (int32_t)value;
+	val->val2 = (((int32_t)(value * 1000) % 1000) * 1000);
+}
+
+static int mc3419_channel_get(const struct device *dev,
+			      enum sensor_channel chan,
+			      struct sensor_value *val)
+{
+	int ret = 0;
+	struct mc3419_driver_data *data = dev->data;
+
+	k_sem_take(&data->sem, K_FOREVER);
+	switch (chan) {
+	case SENSOR_CHAN_ACCEL_X:
+		mc3419_to_sensor_value(data->sensitivity, &data->samples[0], val);
+		break;
+	case SENSOR_CHAN_ACCEL_Y:
+		mc3419_to_sensor_value(data->sensitivity, &data->samples[1], val);
+		break;
+	case SENSOR_CHAN_ACCEL_Z:
+		mc3419_to_sensor_value(data->sensitivity, &data->samples[2], val);
+		break;
+	case SENSOR_CHAN_ACCEL_XYZ:
+		mc3419_to_sensor_value(data->sensitivity, &data->samples[0], &val[0]);
+		mc3419_to_sensor_value(data->sensitivity, &data->samples[1], &val[1]);
+		mc3419_to_sensor_value(data->sensitivity, &data->samples[2], &val[2]);
+		break;
+	default:
+		LOG_ERR("Unsupported channel");
+		ret = -EINVAL;
+	}
+
+	k_sem_give(&data->sem);
+	return ret;
+}
+
+static int mc3419_set_accel_range(const struct device *dev, uint8_t range)
+{
+	int ret = 0;
+	const struct mc3419_config *cfg = dev->config;
+	struct mc3419_driver_data *data = dev->data;
+
+	if (range >= MC3419_ACCL_RANGE_END) {
+		LOG_ERR("Accel resolution is out of range");
+		return -EINVAL;
+	}
+
+	ret = i2c_reg_update_byte_dt(&cfg->i2c, MC3419_REG_RANGE_SELECT_CTRL,
+				     MC3419_RANGE_MASK, range << 4);
+	if (ret < 0) {
+		LOG_ERR("Failed to set resolution (%d)", ret);
+		return ret;
+	}
+
+	data->sensitivity = (double)(mc3419_accel_sense_map[range] *
+				     SENSOR_GRAIN_VALUE);
+
+	return 0;
+}
+
+static int mc3419_set_odr(const struct device *dev,
+			  const struct sensor_value *val)
+{
+	int ret = 0;
+	int data_rate = 0;
+	const struct mc3419_config *cfg = dev->config;
+
+	ret = mc3419_get_odr_value(val->val1, val->val2);
+	if (ret < 0) {
+		LOG_ERR("Failed to get odr value from odr map (%d)", ret);
+		return ret;
+	}
+
+	data_rate = MC3419_BASE_ODR_VAL + ret;
+
+	ret = i2c_reg_write_byte_dt(&cfg->i2c, MC3419_REG_SAMPLE_RATE,
+				    data_rate);
+	if (ret < 0) {
+		LOG_ERR("Failed to set ODR (%d)", ret);
+		return ret;
+	}
+
+	LOG_DBG("Set ODR Rate to 0x%x", data_rate);
+	ret = i2c_reg_write_byte_dt(&cfg->i2c, MC3419_REG_SAMPLE_RATE_2,
+				    CONFIG_MC3419_DECIMATION_RATE);
+	if (ret < 0) {
+		LOG_ERR("Failed to set decimation rate (%d)", ret);
+		return ret;
+	}
+
+	return 0;
+}
+
+#if defined(CONFIG_MC3419_TRIGGER)
+static int mc3419_set_anymotion_threshold(const struct device *dev,
+					  const struct sensor_value *val)
+{
+	int ret = 0;
+	const struct mc3419_config *cfg = dev->config;
+
+	if (val->val1 > MC3419_ANY_MOTION_THRESH_MAX) {
+		return -EINVAL;
+	}
+
+	ret = i2c_burst_write_dt(&cfg->i2c, MC3419_REG_ANY_MOTION_THRES,
```

Comment:
Burst write is problematic and does not work on all devices, it would be better to generate a local buffer and use i2c_write_dt().